src/continuous_agg.c

```diff
@@ -989,3 +999,97 @@ ts_continuous_agg_find_integer_now_func_by_materialization_id(int32 mat_htid)
 	}
 	return par_dim;
 }
+
+TS_FUNCTION_INFO_V1(ts_continuous_agg_watermark);
+
+/*
+ * Get the watermark for a real-time aggregation query on a continuous
+ * aggregate.
+ *
+ * The watermark determines where the materialization ends for a continuous
+ * aggregate. It is used by real-time aggregation as the threshold between the
+ * materialized data and real-time data in the UNION query.
+ *
+ * The watermark is defined as the end of the last (highest) bucket in the
+ * materialized hypertable of a continuous aggregate.
+ *
+ * The materialized hypertable ID is given as input argument.
+ */
+Datum
+ts_continuous_agg_watermark(PG_FUNCTION_ARGS)
+{
+	int32 hyper_id = PG_GETARG_INT32(0);
+	ContinuousAgg *cagg;
+	StringInfo command;
+	Hypertable *ht;
+	Dimension *dim;
+	Datum maxdat;
+	bool max_isnull;
+	int64 watermark;
+	Oid timetype;
+	AclResult aclresult;
+	int res;
+
+	if (PG_ARGISNULL(0))
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+				 errmsg("invalid continuous aggregate hypertable")));
+
+	cagg = ts_continuous_agg_find_by_mat_hypertable_id(hyper_id);
+
+	if (NULL == cagg)
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+				 errmsg("%d is not a materialized hypertable", hyper_id)));
+
+	/* Preemptive permission check to ensure the function complains about lack
+	 * of permissions on the cagg rather than the materialized hypertable */
+	aclresult = pg_class_aclcheck(cagg->relid, GetUserId(), ACL_SELECT);
+	aclcheck_error(aclresult, OBJECT_MATVIEW, get_rel_name(cagg->relid));
+
+	ht = ts_hypertable_get_by_id(hyper_id);
+	Assert(NULL != ht);
+	dim = hyperspace_get_open_dimension(ht->space, 0);
+	timetype = ts_dimension_get_partition_type(dim);
+
+	res = SPI_connect();
+
+	if (res != SPI_OK_CONNECT)
```

Comment:
This is overwritten later, so it could hide a bug where the `rel` is not initialized properly.
```suggestion
	if (SPI_connect() != SPI_OK_CONNECT)
```